src/FEValues/face_values.jl

```diff
@@ -0,0 +1,274 @@
+# Defines FaceScalarValues and FaceVectorValues and common methods
+"""
+A `FaceValues` object facilitates the process of evaluating values of shape functions, gradients of shape functions,
+values of nodal functions, gradients and divergences of nodal functions etc. on the faces of finite elements. There are
+two different types of `FaceValues`: `FaceScalarValues` and `FaceVectorValues`. As the names suggest,
+`FaceScalarValues` utilizes scalar shape functions and `FaceVectorValues` utilizes vectorial shape functions.
+For a scalar field, the `FaceScalarValues` type should be used. For vector field, both subtypes can be used.
+
+**Constructors:**
+
+Note: The quadrature rule for the face should be given with one dimension lower. I.e. for a 3D case, the quadrature rule
+should be in 2D.
+
+```julia
+FaceScalarValues([::Type{T}], quad_rule::QuadratureRule, func_interpol::Interpolation, [geom_interpol::Interpolation])
+FaceVectorValues([::Type{T}], quad_rule::QuadratureRule, func_interpol::Interpolation, [geom_interpol::Interpolation])
+```
+
+**Arguments:**
+
+* `T`: an optional argument to determine the type the internal data is stored as.
+* `quad_rule`: an instance of a [`QuadratureRule`](@ref)
+* `func_interpol`: an instance of an [`Interpolation`](@ref) used to interpolate the approximated function
+* `geom_interpol`: an optional instance of an [`Interpolation`](@ref) which is used to interpolate the geometry
+
+**Common methods:**
+
+* [`reinit!`](@ref)
+* [`getfacenumber`](@ref)
+* [`getnquadpoints`](@ref)
+* [`getquadrule`](@ref)
+* [`getfunctioninterpolation`](@ref)
+* [`getgeometryinterpolation`](@ref)
+* [`getdetJdV`](@ref)
+
+* [`shape_value`](@ref)
+* [`shape_gradient`](@ref)
+* [`shape_symmetric_gradient`](@ref)
+* [`shape_divergence`](@ref)
+
+* [`function_value`](@ref)
+* [`function_gradient`](@ref)
+* [`function_symmetric_gradient`](@ref)
+* [`function_divergence`](@ref)
+* [`spatial_coordinate`](@ref)
+"""
+FaceValues
+
+# FaceScalarValues
+immutable FaceScalarValues{dim, T <: Real, refshape <: AbstractRefShape} <: FaceValues{dim, T, refshape}
+    N::Array{T, 3}
+    dNdx::Array{Vec{dim, T}, 3}
+    dNdξ::Array{Vec{dim, T}, 3}
+    detJdV::Matrix{T}
+    normals::Vector{Vec{dim, T}}
+    M::Array{T, 3}
+    dMdξ::Array{Vec{dim, T}, 3}
+    qr_weights::Vector{T}
+    current_face::ScalarWrapper{Int}
```

Comment:
Actually, `Ref` is not an leaf type so using it created type instabilities. Created a new (mutable) type that just wraps a value.